tutorials/tagger/basic_allennlp_annotated.py

```diff
@@ -0,0 +1,204 @@
+token_embedding = Embedding(num_embeddings=vocab.get_vocab_size('tokens'),
+                            embedding_dim=EMBEDDING_DIM)
+word_embeddings = BasicTextFieldEmbedder({"tokens": token_embedding})
+#### We next need to specify the sequence encoder. The need for <code>PytorchSeq2SeqWrapper</code> here is slightly unfortunate (and later we'll show you how not to have to worry about it) but it's required to add some extra functionality (and a cleaner interface) to the built in PyTorch module. In AllenNLP we do everything batch first, so we specify that as well.
+lstm = PytorchSeq2SeqWrapper(torch.nn.LSTM(EMBEDDING_DIM, HIDDEN_DIM, batch_first=True))
+
+#### Finally, we can instantiate the model.
+model = LstmTagger(word_embeddings, lstm, vocab)
+
+#### Now we're ready to train the model. The first thing we'll need is an optimizer. We can just use PyTorch's stochastic gradient descent.
+optimizer = optim.SGD(model.parameters(), lr=0.1)
+
+#### And we need a <code>DataIterator</code> that handles batching for our datasets.
+iterator = BasicIterator(batch_size=2)
+#### We also specify that the iterator should make sure its instances are indexed using our vocabulary.
+iterator.index_with(vocab)
+
+#### Finally, we can instantiate our <code>Trainer</code> and run it. Here we tell it to run for 1000 epochs and to stop training early if it ever spends 10 epochs without the validation loss improving.
+trainer = Trainer(model=model,
+                  optimizer=optimizer,
+                  iterator=iterator,
+                  train_dataset=train_dataset,
+                  validation_dataset=validation_dataset,
+                  patience=10,
+                  num_epochs=1000)
+
+#### When we launch it it will print a progress bar for each epoch that includes both the "loss" and the "accuracy" metric. If our model is good, the loss should go down and the accuracy up as we train.
+trainer.train()
+
+#### As in the original PyTorch tutorial, we'd like to look at the predictions our model generates. AllenNLP contains a <code>Predictor</code> abstraction that takes inputs, converts them to instances, feeds them through your model, and returns JSON-serializable results. Often you'd need to implement your own Predictor, but AllenNLP already has a <code>SentenceTaggerPredictor</code> that works perfectly here, so we can use it. It requires our model (for making predictions) and a dataset reader (for creating instances).
+predictor = SentenceTaggerPredictor(model, dataset_reader=reader)
+#### It has a <code>predict</code> method that just needs a sentence and returns (a JSON-serializable version of) the output dict from forward.  Here "tag_scores" will be a (5, 3) array of logits, corresponding to the 3 possible tag for each of the 5 words.
```

Comment:
3 possible tags

Reply:
And when I first read this I was confused at the mismatch between "tag_logits" and "tag_scores".  Also, "tag_scores" in the annotation should probably be in a <code> block.